### PR TITLE
service functions & mock data

### PR DESCRIPTION
**Pull request by pldespaigne**

### CODE REVIEW

File: libs/organization/src/lib/+state/organization.model.ts [hunk 1]
```diff
@@ -1,9 +1,27 @@
 export interface OrgMember {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.model.ts]: Call it OrganizationMember

File: libs/organization/src/lib/+state/organization.model.ts [hunk 2]
```diff
@@ -1,9 +1,27 @@
...
+  avatar?: string;
   uid: string;
+  name?: string;
   email: string;
   roles: string[];
 }
 
+export interface OrgOperation {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.model.ts]: OrganizationOperation

File: libs/organization/src/lib/+state/organization.model.ts [hunk 3]
```diff
@@ -1,9 +1,27 @@
...
+  id: string;
+  name: string;
+	quorum: number;
+	members: OrgMember[]
+}
+
+export interface OrgAction {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.model.ts]: OrganizationAction

File: libs/organization/src/lib/+state/organization.service.ts [hunk 4]
```diff
@@ -103,4 +103,122 @@ export class OrganizationService {
...
+    newOperations.push(newOperation); // we could not use `actions.push(newAction)` direclty otherwise the action will have been duplicated
+
+    try {
+      // send tx to the org smart-contract and wait for result // TODO replace with the real implemntation : issue 676
+
+      // update the store
+      this.store.update(state => {
+        return {
+          ...state, // keep everything of the state
+          org: { // update only the org
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.service.ts]: Too many comments ^^.
```typescript
this.store.update(state => ({
  ...state, // keep everything of the state
 org: { ...state.org, operations: newOperations }  // update the actions array
}))
```

File: libs/organization/src/lib/+state/organization.service.ts [hunk 5]
```diff
@@ -103,4 +103,122 @@ export class OrganizationService {
...
   }
+
+  //-------------------------------------------------
+  //            BLOCKCHAIN PART OF ORGS
+  //-------------------------------------------------
+
+  /** create a newOperation, or update it if it already exists */
+  private async upsertOperation(newOperation: OrgOperation) {
+    const { operations } = this.query.getValue().org // get every actions
+    const newOperations = operations.filter(currentOperation => currentOperation.id !== newOperation.id); // get all actions except the one we want to upsert
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.service.ts]: Avoid mutation when possible : 
```typescript
const newOperations = [
  ...operations.filter(currentOperation => currentOperation.id !== newOperation.id), all actions except the one we want to upsert
  newOperation // updated action
]
```

File: libs/organization/src/lib/+state/organization.service.ts [hunk 6]
```diff
@@ -103,4 +103,122 @@ export class OrganizationService {
...
+      });
+    } catch(err) {
+      console.error('The transaction has failed :', err); // TODO better error handling : issue 671
+    }
+  }
+
+  updateOperationQuorum(id: string, newQuorum: number) {
+    const operation = this.query.getOperationById(id);
+    if(!operation) throw new Error('This operation doesn\'t exists');
+    const newOperation = {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.service.ts]: I think here the intermediary variable is too much

File: libs/organization/src/lib/+state/organization.service.ts [hunk 7]
```diff
@@ -103,4 +103,122 @@ export class OrganizationService {
...
+      quorum: newQuorum,
+    };
+    this.upsertOperation(newOperation);
+  }
+
+  addOperationMember(id: string, newMember: OrgMember) {
+    const operation = this.query.getOperationById(id);
+    if(!operation) throw new Error('This operation doesn\'t exists');
+    
+    const [memberExists] = operation.members.filter(member => member.uid === newMember.uid);
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.service.ts]: Use find when you're look for one: 
```typescript
const memberExists = operation.members.find(member => member.uid === newMember.uid);
```
BUT in your case you want to check if the array `includes` the member
```typescript
const memberExists = operation.members.includes(member => member.uid === newMember.uid); // boolean
```

File: libs/organization/src/lib/+state/organization.service.ts [hunk 8]
```diff
@@ -103,4 +103,122 @@ export class OrganizationService {
...
+    if (!!memberExists) throw new Error('This member is already a signer of this operation');
+
+    const newOperation = {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.service.ts]: same as above

File: libs/organization/src/lib/+state/organization.service.ts [hunk 9]
```diff
@@ -103,4 +103,122 @@ export class OrganizationService {
...
+    };
+    
+    this.upsertOperation(newOperation);
+  }
+
+  removeOperationMember(id: string, memberToRemove: OrgMember) {
+    const operation = this.query.getOperationById(id);
+    if(!operation) throw new Error('This operation doesn\'t exists');
+    
+    const newMembersList = operation.members.filter(member => member.uid !== memberToRemove.uid);
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.service.ts]: call it members : 
```typescript
const members = operation.members.filter(member => member.uid !== memberToRemove.uid);
this.upsertOperation({...operation, members });
```

File: libs/organization/src/lib/+state/organization.service.ts (hunk 7 above)
Comment (pldespaigne) [reply libs/organization/src/lib/+state/organization.service.ts]: `includes()` cannot take an arrow function as a parameters, it accept only a value, and perform a value equality test, witch will not work on objects.
BUT in my case I think I can use `some() => boolean`
```typescript
const memberExists = operation.members.some(member => member.uid === newMember.uid);
```

What do you think ?

Comment (GrandSchtroumpf) [reply libs/organization/src/lib/+state/organization.service.ts]: Yes `some` is good.

File: libs/organization/src/lib/+state/organization.service.ts [hunk 10]
```diff
@@ -103,4 +103,115 @@ export class OrganizationService {
...
+    if(!operation) throw new Error('This operation doesn\'t exists');
+    
+    const members = operation.members.filter(member => member.uid !== memberToRemove.uid);
+    const newOperation = { ...operation, members };
+    
+    this.upsertOperation(newOperation);
+  }
+
+  // TODO REMOVE THIS ASAP : issue 676
+  public instantiateMockData() {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.service.ts]: Put the mocked value inside a .mock.ts file inside +state folder.

**Pull request merged**